add detailed error messages to @ogrerr

mattwigway opened the pull request:
The @ogrerr macro just printed whatever was passed to it in the source, which was often unhelpful (for instance, in `setfeature!`, "Failed to set feature." This change prints a human-readable message based on the OGR error code, in addition to the message passed to ogrerr.

yeesian commented on src/utils.jl:
```diff
@@ -184,7 +184,30 @@ end
-            error($message)
+            # find the detailed error message
+            if $(esc(code)) == GDAL.OGRERR_NOT_ENOUGH_DATA
+                detailmsg = "Not enough data."
```
Can we make it `detailmsg = if ...` so that we are not defining new variables inside the if statement?

mattwigway replied:
I always forget you can do this in Julia - done.

yeesian reviewed (changes requested):
Thank you, this is great :)

Can we add some unit tests to show that the macro is working as intended? Maybe something in https://github.com/yeesian/ArchGDAL.jl/blob/master/test/test_utils.jl along the lines of 
```julia
using Test
import GDAL
import ArchGDAL;
const AG = ArchGDAL;

eval_ogrerr(code, message) = AG.@ogrerr(code, message)

@testset "test_utils.jl" begin
    @testset "metadataitem" begin
        driver = AG.getdriver("DERIVED")
        @test AG.metadataitem(driver, "DMD_EXTENSIONS") == ""
        driver = AG.getdriver("GTiff")
        @test AG.metadataitem(driver, "DMD_EXTENSIONS") == "tif tiff"
    end

    @testset "OGR Errors" begin
        @test isnothing(eval_ogrerr(GDAL.OGRERR_NONE, "message"))
        @test_throws ErrorException eval_ogrerr(GDAL.OGRERR_NOT_ENOUGH_DATA, "error message")
        @test_throws ErrorException eval_ogrerr(GDAL.OGRERR_NOT_ENOUGH_MEMORY, "error message")
        @test_throws ErrorException eval_ogrerr(GDAL.OGRERR_UNSUPPORTED_GEOMETRY_TYPE, "error message")
        @test_throws ErrorException eval_ogrerr(GDAL.OGRERR_UNSUPPORTED_OPERATION, "error message")
        @test_throws ErrorException eval_ogrerr(GDAL.OGRERR_CORRUPT_DATA, "error message")
        @test_throws ErrorException eval_ogrerr(GDAL.OGRERR_FAILURE, "error message")
        @test_throws ErrorException eval_ogrerr(GDAL.OGRERR_UNSUPPORTED_SRS, "error message")
        @test_throws ErrorException eval_ogrerr(GDAL.OGRERR_INVALID_HANDLE, "error message")
        @test_throws ErrorException eval_ogrerr(GDAL.OGRERR_NON_EXISTING_FEATURE, "error message")
    end
end
```

mattwigway commented:
Okay, addressed the style comment, and added a test that checks that errors are thrown, and that the error messages are correct. Let me know if there's anything else I can do!
